### ISSUE TITLE
Swarm tasks inconsistently stuck in New state on service creation

### ISSUE THREAD

**Issue by mattlqx** [moby/moby#35849]
**Description**

Inconsistently, some services will fail to assign some or all of their tasks. For example:

```
root@qa-swarm-manager1:~# docker service ps uname-echo
ID                  NAME                IMAGE               NODE                                    DESIRED STATE       CURRENT STATE           ERROR               PORTS
1d4qk1owa4q4        uname-echo.1        busybox:latest      qa-swarm-worker1                   Running             Running 5 minutes ago                       
qmyqh3t36qmz        uname-echo.2        busybox:latest                                              Running             New 5 minutes ago                           
ttxzdobi8xu5        uname-echo.3        busybox:latest                                              Running             New 5 minutes ago  
```

Where there are three worker nodes in total. I have a manager that the service was created on and a worker that the task is not running on running at debug log level and no relevant log messages to the service appear to be logged aside from this on worker2 to learn about the task on worker1 (I presume).

```
Dec 20 08:10:01 qa-swarm-worker2 dockerd[15414]: time="2017-12-20T08:10:01.242685624-08:00" level=debug msg="handleEpTableEvent ADD 9a34dae4bd58a72b26d1bf2dd996058ae7776f76d6d72bfc1078fcf1a6ce13cb R:{uname-echo.1.1d4qk1owa4q4ngr9gww0y2yo7 uname-echo pheha2c4cvxmc6xoocqp7b6n8 10
.255.0.104 10.255.0.105 [&PortConfig{Name:,Protocol:TCP,TargetPort:2000,PublishedPort:2000,}] [] [184055cdffca]}"
Dec 20 08:10:01 qa-swarm-worker2 dockerd[15414]: time="2017-12-20T08:10:01.242832003-08:00" level=debug msg="addServiceBinding from handleEpTableEvent START for uname-echo 9a34dae4bd58a72b26d1bf2dd996058ae7776f76d6d72bfc1078fcf1a6ce13cb p:0xc422248000 nid:1sxb29yly08nza2lca7k85
8zc skey:{pheha2c4cvxmc6xoocqp7b6n8 2000:2000/TCP}"
Dec 20 08:10:01 qa-swarm-worker2 dockerd[15414]: time="2017-12-20T08:10:01.243277282-08:00" level=debug msg="/sbin/iptables, [--wait -t nat -nL DOCKER-INGRESS]"
Dec 20 08:10:01 qa-swarm-worker2 dockerd[15414]: time="2017-12-20T08:10:01.253861877-08:00" level=debug msg="/sbin/iptables, [--wait -t filter -nL DOCKER-INGRESS]"
Dec 20 08:10:01 qa-swarm-worker2 dockerd[15414]: time="2017-12-20T08:10:01.263331739-08:00" level=debug msg="/sbin/iptables, [--wait -t nat -C DOCKER-INGRESS -j RETURN]"
Dec 20 08:10:01 qa-swarm-worker2 dockerd[15414]: time="2017-12-20T08:10:01.274421025-08:00" level=debug msg="/sbin/iptables, [--wait -t filter -C DOCKER-INGRESS -j RETURN]"
Dec 20 08:10:01 qa-swarm-worker2 dockerd[15414]: time="2017-12-20T08:10:01.281857420-08:00" level=debug msg="/sbin/iptables, [--wait -t nat -C OUTPUT -m addrtype --dst-type LOCAL -j DOCKER-INGRESS]"
Dec 20 08:10:01 qa-swarm-worker2 dockerd[15414]: time="2017-12-20T08:10:01.289721075-08:00" level=debug msg="/sbin/iptables, [--wait -t nat -C PREROUTING -m addrtype --dst-type LOCAL -j DOCKER-INGRESS]"
Dec 20 08:10:01 qa-swarm-worker2 dockerd[15414]: time="2017-12-20T08:10:01.299417738-08:00" level=debug msg="/sbin/iptables, [--wait -t filter -C FORWARD -j DOCKER-INGRESS]"
Dec 20 08:10:01 qa-swarm-worker2 dockerd[15414]: time="2017-12-20T08:10:01.308027196-08:00" level=debug msg="/sbin/iptables, [--wait -t nat -C POSTROUTING -m addrtype --src-type LOCAL -o docker_gwbridge -j MASQUERADE]"
Dec 20 08:10:01 qa-swarm-worker2 dockerd[15414]: time="2017-12-20T08:10:01.318947307-08:00" level=debug msg="/sbin/iptables, [--wait -t nat -nL DOCKER-INGRESS]"
Dec 20 08:10:01 qa-swarm-worker2 dockerd[15414]: time="2017-12-20T08:10:01.331459121-08:00" level=debug msg="/sbin/iptables, [--wait -t nat -I DOCKER-INGRESS -p tcp --dport 2000 -j DNAT --to-destination 172.18.0.2:2000]"
Dec 20 08:10:01 qa-swarm-worker2 dockerd[15414]: time="2017-12-20T08:10:01.341649297-08:00" level=debug msg="/sbin/iptables, [--wait -I DOCKER-INGRESS -m state -p tcp --sport 2000 --state ESTABLISHED,RELATED -j ACCEPT]"
Dec 20 08:10:01 qa-swarm-worker2 dockerd[15414]: time="2017-12-20T08:10:01.353493423-08:00" level=debug msg="/sbin/iptables, [--wait -I DOCKER-INGRESS -p tcp --dport 2000 -j ACCEPT]"
Dec 20 08:10:01 qa-swarm-worker2 dockerd[15414]: time="2017-12-20T08:10:01.364219644-08:00" level=debug msg="Creating service for vip 10.255.0.104 fwMark 258 ingressPorts []*libnetwork.PortConfig{&libnetwork.PortConfig{Name: \"\",\nProtocol: 0,\nTargetPort: 0x7d0,\nPublishedPor
t: 0x7d0,\n}} in sbox ingress (ingress)"
Dec 20 08:10:01 qa-swarm-worker2 dockerd[15414]: time="2017-12-20T08:10:01.466581471-08:00" level=debug msg="addEndpointNameResolution 9a34dae4bd58a72b26d1bf2dd996058ae7776f76d6d72bfc1078fcf1a6ce13cb uname-echo add_service:true sAliases:[] tAliases:[184055cdffca]"
Dec 20 08:10:01 qa-swarm-worker2 dockerd[15414]: time="2017-12-20T08:10:01.466685837-08:00" level=debug msg="addContainerNameResolution 9a34dae4bd58a72b26d1bf2dd996058ae7776f76d6d72bfc1078fcf1a6ce13cb uname-echo.1.1d4qk1owa4q4ngr9gww0y2yo7"
Dec 20 08:10:01 qa-swarm-worker2 dockerd[15414]: time="2017-12-20T08:10:01.466712490-08:00" level=debug msg="addServiceBinding from handleEpTableEvent END for uname-echo 9a34dae4bd58a72b26d1bf2dd996058ae7776f76d6d72bfc1078fcf1a6ce13cb"
```

This is the whole list of services on the cluster:
```
ID                  NAME                MODE                REPLICAS            IMAGE                                               PORTS
iyebje9w7uli        es-proxy            replicated          2/2                 h3nrik/nginx-ldap:latest                            *:15002->443/tcp
m3orsxa9hwn0        nginx               global              3/3                 nexus:5000/nginx-jdomain:latest   
e9r10puh8vqi        uname-echo          replicated          1/3                 busybox:latest                                      *:2000->2000/tcp
```

Depending on the order in which the services are created, it may not always be the same service that hits this. I'm not seeing any indication of why it's failing to progress from New state on these tasks. The uname-echo service has an ingress network port mapped, but is not a member of any other networks.

**Steps to reproduce the issue:**
1. Create service on swarm with ingress port mapped.

**Describe the results you received:**

Some tasks never get assigned to workers and get stuck in "New" status.

**Describe the results you expected:**

All tasks to be assigned to workers.

**Additional information you deem important (e.g. issue happens only occasionally):**

uname-echo service inspect: https://gist.github.com/mattlqx/cbf38fef481ffe2d399119c8342511b7
ingress network inspect: https://gist.github.com/mattlqx/f82ba1c9fe64128b96a4a8d75dacd01f

**Output of `docker version`:**

```
Docker version 17.11.0-ce, build 1caf76c
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 17.11.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: ncmrtjzjdbn2hrxo02lfk2y9i
 Is Manager: true
 ClusterID: 29bbrozr1vt4vtfuwpxawcijb
 Managers: 3
 Nodes: 6
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: 172.16.7.224
 Manager Addresses:
  172.16.7.224:2377
  172.16.7.225:2377
  172.16.7.226:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 992280e8e265f491f7a624ab82f3e238be086e49
runc version: 0351df1c5a66838d0c392b4ac4cf9450de844e2d
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-98-generic
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.953GiB
Name: qa-swarm-manager1
ID: HHRF:ZAK7:CTGM:FRIQ:FMGP:CMXE:DTWF:VOTY:PCM5:LOYA:UIG7:U3RP
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 45
 Goroutines: 139
 System Time: 2017-12-20T08:35:15.192929619-08:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
vSphere instances

**Comment by mattlqx** [comment]
I decided to try upgrading the kernels to 4.13.0 and rebooting the swarm sequentially with the managers first. When manager3 which was the leader was rebooted and a new leader took over, the tasks for the service got assigned and everything looks "fully converged". I'm not sure if the kernel had any effect, but I think it was more likely the act of failing over the leader that unstuck it.

**Comment by cima** [comment]
Are there gonna be any response from docker team? We have experienced the same issue while having 14 nodes, about 8 overlay networks and groups services that only differs in network presence while having the same docker image. Some of them started, some of them remained in new state. Even after destroying the whole stack and recreating it from scratch, the same services stuck in new state again.

We are using virtual machines and scale sets on Azure and some services are created programmatically. Docker 17.12.

**Comment by afiestas** [comment]
We were just having the same behavior, we fixed it by rebooting 2 of our 6 managers. Any extra information required to debug this issue we should be able to provide.

**Comment by cima** [comment]
After few hours we have been able to start the rest of the containers. Our problem was that some services crashed (our internal problem, not related to docker) and as the result tasks remained in history. For that purpose, container is retained on node and it keeps the IP address. We only had /24 network (256 addresses) and suddenly all were taken by both running and exited containers. We have make the network bigger /16 and now we have all containers running. We have also lowered the task history length. I filled bug for inconsistency/ambiguity in documentation of docker [Inconsistent overlay network size limits](https://github.com/docker/docker.github.io/issues/5853).

**Comment by afiestas** [comment]
@cima it seems that the same was happening to us, that was spot on!

How did you debug this? Were there any interesting logs around? We found none :/

Thanks!

**Comment by cima** [comment]
Wa have listed all tasks for our docker stack. There were more than 256 of them. Then we inspected several nodes and by listing using `docker ps -a` we saw exited containers. Such container retains IP address. I reconfigured the default network to be /16 and re-created entire stack. Everything started like a charm. We haven't look into master's linux service of docker daemon. Retrospectively I have look into daemon logs on both leading masters as well as in `dmesg` and both showed no entry that would lead us to our solution. So we figured out the situation just based on knowledge of documentation, networking and actual container situation on each node.

**Comment by afiestas** [comment]
@cima So perhaps this calls for opening a new issue to add some logs when this issue is hit plus some information in docker service ps.

**Comment by cima** [comment]
I agree. I spent some extra time to prepare short example. We have 1 master and 5 workers.
Suppose yaml file:
```
version: '3.3'

networks:
  default:
    ipam:
      config:
        - subnet: 10.135.0.0/26
        
services:
  networking-service:
    image: "jamesw/http-echo-server"
    deploy:
      mode: replicated
      replicas: 1
    ports:
      - "8080:8080"
    environment:
      - text=Gizmo likes pizza
      - listen=:8080
```
Create stack by calling
```
docker stack deploy -c docker-compose.yml gizmobox
```
See service status
```
opijyztj33xp    gizmobox_networking-service   replicated   1/1   jamesw/http-echo-server:latest
```
Remove the stack
```
docker stack rm gizmobox
```
Now modify the network range to have network only **/30**
```
version: '3.3'

networks:
  default:
    ipam:
      config:
        - subnet: 10.135.0.0/30
        
services:
  networking-service:
    image: "jamesw/http-echo-server"
    deploy:
      mode: replicated
      replicas: 1
    ports:
      - "8080:8080"
    environment:
      - text=Gizmo likes pizza
      - listen=:8080
```
build stack again
```
docker stack deploy -c docker-compose.yml gizmobox
```
See services
```
l7lzas9br18y   gizmobox_networking-service   replicated   0/1   jamesw/http-echo-server:latest
```
Bummer. Lets list some logs (command might differ based on distribution of linux)
```
sudo journalctl -fu docker.service
```
and delight with and output
```
-- Logs begin at Tue 2018-01-23 19:07:34 UTC. --
Jan 30 08:13:13 master-vm dockerd[867]: time="2018-01-30T08:13:13.243542757Z" level=error msg="Failed to allocate network resources for node ru2fuz57o6p8ufj7q629haoul" error="could not find an available IP" module=node node.id=zjmt9ej7uzlmk37l35i37lonl
Jan 30 08:13:13 master-vm dockerd[867]: time="2018-01-30T08:13:13.243593857Z" level=error msg="Failed to allocate network resources for node yehx4ucbx4wxnci6b5864fulb" error="could not find an available IP" module=node node.id=zjmt9ej7uzlmk37l35i37lonl
Jan 30 08:13:13 master-vm dockerd[867]: time="2018-01-30T08:13:13.243649357Z" level=error msg="Failed to allocate network resources for node zjmt9ej7uzlmk37l35i37lonl" error="could not find an available IP" module=node node.id=zjmt9ej7uzlmk37l35i37lonl
Jan 30 08:13:14 master-vm dockerd[867]: time="2018-01-30T08:13:14.832076366Z" level=error msg="Failed allocation for service l7lzas9br18ytp36o962u8imu" error="could not find an available IP while allocating VIP" module=node node.id=zjmt9ej7uzlmk37l35i37lonl
Jan 30 08:13:14 master-vm dockerd[867]: time="2018-01-30T08:13:14.837551065Z" level=error msg="task allocation failure" error="service l7lzas9br18ytp36o962u8imu to which this task u1j2isjr01oc1wtx4ca762qz7 belongs has pending allocations" module=node node.id=zjmt9ej7uzlmk37l35i37lonl
```

To be fair I stated above that I didn't saw any logs but in fact they just rolled out. There is a log output but in some distribution quite volatile and not very explaining for users with low to intermediate networking knowledge.

Apparently even worker nodes consume some IP addresses.

My suggestion is to advertise this error along with a task, not deeply in a log. PENDING state would be more helpful instead of NEW which everyone treats like _"Wait for it. It will start loading soon. NEW state is a good one."_ Lack of IP addresses is similar to lack of resources and requires interception of operator either by change of the network or forcible removal of exited containers that retains IP addresses.

**Comment by shabbirkagalwala** [comment]
I have been stuck on the same issue too, all of a sudden today we started seeing this issue when starting a new service. We have a cluster of 9 nodes and 2 overlays we use. More than a 100 but less than 150 services and for some reason the task gets stuck on the "New" state. I don't see any errors like "could not find an available IP". Could someone please help what could be done to resolve the issue?

docker network inspect --verbose --format  '{{range .Services}}{{printf "%s\n" .VIP}}{{range .Tasks}}{{printf "%s\n" .EndpointIP}}{{end}}{{end}}' <network_name> |grep -v '^$' |wc -l

Shows the number of IPs allocated on the network very high than the number of tasks in total, also no dead containers on the nodes

**Comment by cima** [comment]
Are you sure there are no hanging containers? Swarm keeps containers for tasks that have already ended. We had exactly this problem. Writing `docker ps` showed few container at each node. But writing `docker ps -a` revealed some stopped container. We rebooted all nodes, but later I tried to lower the task history limit and it pruned dead containers. On manager node execute following command

`docker swarm update --task-history-limit=1`

See [How to clear Docker task history](https://stackoverflow.com/questions/42364695/how-to-clear-docker-task-history) for details.

**Comment by rdehouss** [comment]
I'm having the exact same issue

It's stuck on all my nodes
![image](https://user-images.githubusercontent.com/461529/91083194-28851480-e64a-11ea-914a-ffdf12775982.png)

**Output of `docker version`:**
```
Client: Docker Engine - Community
 Version:           19.03.12
 API version:       1.40
 Go version:        go1.13.10
 Git commit:        48a66213fe
 Built:             Mon Jun 22 15:45:50 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.12
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.10
  Git commit:       48a66213fe
  Built:            Mon Jun 22 15:44:21 2020
  OS/Arch:          linux/amd64
  Experimental:     true
 containerd:
  Version:          1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Output of `docker info`:**
```
Client:
 Debug Mode: false

Server:
 Containers: 10
  Running: 10
  Paused: 0
  Stopped: 0
 Images: 13
 Server Version: 19.03.12
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: active
  NodeID: ecoktc7wfqgtrj8d5ztom7zhn
  Is Manager: false
  Node Address: 172.16.0.121
  Manager Addresses:
   172.16.0.113:2377
   172.16.0.128:2377
   172.16.0.87:2377
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 4.19.0-5-cloud-amd64
 Operating System: Debian GNU/Linux 10 (buster)
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 14.34GiB
 Name: common-002
 ID: 4O6P:KTMM:IHJR:RUFS:QPO5:62SE:WGKH:5ATU:MUYL:AIM7:C3DH:QDBZ
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: true
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No swap limit support
```

**Comment by Danylov-Mykola** [comment]
I have the same issue. Need help.

**Comment by cima** [comment]
@rdehouss & @Danylov-Mykola I suggest to try `docker ps -a` because I bet there are dead containers that consumes IP addresses needed by new instances. Se my previous comments for reasoning.

Maybe it is inappropriate to state here but I'd recommend you to switch to Kubernetes. That is how we have resolved this issue for good in the company I was working at the time.

**Comment by mattlqx** [comment]
I honestly did the same. We've been running k8s for 2 years now and I've haven't ever been blocked by issues like this.

**Comment by seb350** [comment]
Had the same problem and in my case I had the same "Subnet": "10.0.0.0/24" for 2 different networks. 

The first one was being created properly with this subnet somehow by default and for the second one I was manually specifying this same subnet in the .yml file. I can't understand how I managed to specify the exact same subnet to be used.

Docker doesn't throw any errors regarding this, it just won't create the network properly and the containers won't be added to it therefore being stuck.

Maybe it will help someone else even tho it won't cover all cases.
